entity-framework/core/modeling/value-conversions.md

```diff
@@ -2,145 +2,746 @@
+
+### Use ulong for timestamp/rowversion
+
+SQL Server supports automatic [optimistic concurrency](xref:core/saving/concurrency) using [8-byte binary rowversion/timestamp columns](/sql/t-sql/data-types/rowversion-transact-sql). These are always read from and written to the database using an 8-byte array. However, byte arrays are a mutable reference type, which makes them someone painful to deal with. Value converters allow the rowversion to instead be mapped to a `ulong` property, which is much more appropriate and easy to use than the byte array. For example, consider a `Blog` entity with a ulong concurrency token:
+
+<!--
+        public class Blog
+        {
+            public int Id { get; set; }
+            public string Name { get; set; }
+            public ulong Version { get; set; }
+        }
+-->
+[!code-csharp[ULongConcurrencyModel](../../../samples/core/Modeling/ValueConversions/ULongConcurrency.cs?name=ULongConcurrencyModel)]
+
+This can be mapped to a SQL server rowversion column using a value converter:
+
+<!--
+                modelBuilder.Entity<Blog>()
+                    .Property(e => e.Version)
+                    .IsRowVersion()
+                    .HasConversion<byte[]>();
+-->
+[!code-csharp[ConfigureULongConcurrency](../../../samples/core/Modeling/ValueConversions/ULongConcurrency.cs?name=ConfigureULongConcurrency)]
+
+### Specify the DateTime.Kind when reading dates
+
+SQL Server discards the <xref:System.DateTime.Kind%2A?displayProperty=nameWithType> flag when storing a <xref:System.DateTime> as a [datetime](/sql/t-sql/data-types/datetime-transact-sql) or [datetime2](/sql/t-sql/data-types/datetime2-transact-sql). This means that DateTime values coming back from the database always have <xref:System.DateTimeKind> of `Unspecified`.
+
+Value converters can be used in two ways to deal with this. First, EF Core has a value converter that creates an 8-byte opaque value which preserves the `Kind` flag. For example:
+
+<!--
+                modelBuilder.Entity<Post>()
+                    .Property(e => e.PostedOn)
+                    .HasConversion<long>();
+-->
+[!code-csharp[ConfigurePreserveDateTimeKind1](../../../samples/core/Modeling/ValueConversions/PreserveDateTimeKind.cs?name=ConfigurePreserveDateTimeKind1)]
+
+This allows DateTime values with different `Kind` flags to be mixed in the database.
+
+The problem with this approach is that the database no longer has recognizable `datetime` or `datetime2` columns. So instead it is common to always store UTC time (or, less commonly, always local time) and then either ignore the `Kind` flag or set it to the appropriate value using a value converter. For example, the converter below ensures that the `DateTime` value read from the database will have the <xref:System.DateTimeKind> `UTC` kind:
+
+<!--
+                modelBuilder.Entity<Post>()
```

Comment:
I understand this example is only about correcting the Kind for values coming back from the database... But should we also show conversion to UTC when *sending* to the database (e.g. in case a DateTime with Kind=Local is assigned to the entity property)? This would mean calling [DateTime.ToUniversalTime](https://docs.microsoft.com/en-us/dotnet/api/system.datetime.touniversaltime?view=net-5.0#System_DateTime_ToUniversalTime) in the model->provider conversion side.

Am thinking about unsuspecting users copy-pasting this and thinking they have a full, UTC-everywhere solution.

Reply:
Interesting idea; I'll add something.